File: app/src/main/java/com/siju/acexplorer/home/view/HomeLibrary.java
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.home.view;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.CardView;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.View;
import android.widget.ImageButton;

import com.siju.acexplorer.R;
import com.siju.acexplorer.analytics.Analytics;
import com.siju.acexplorer.home.LibrarySortActivity;
import com.siju.acexplorer.home.model.LibrarySortModel;
import com.siju.acexplorer.home.types.HomeLibraryInfo;
import com.siju.acexplorer.model.FileInfo;
import com.siju.acexplorer.model.StorageUtils;
import com.siju.acexplorer.model.groups.Category;
import com.siju.acexplorer.storage.view.custom.helper.SimpleItemTouchHelperCallback;
import com.siju.acexplorer.theme.Theme;
import com.siju.acexplorer.utils.ConfigurationHelper;

import java.util.ArrayList;
import java.util.List;

import static com.siju.acexplorer.model.groups.Category.ADD;
import static com.siju.acexplorer.model.groups.Category.DOWNLOADS;
import static com.siju.acexplorer.model.groups.Category.FAVORITES;

/**
 * Created by Siju on 03 September,2017
 */
class HomeLibrary {

    private final String TAG                  = this.getClass().getSimpleName();
    static final  int    LIBSORT_REQUEST_CODE = 1000;

    private HomeUiView     homeUiView;
    private RecyclerView   libraryList;
    private CardView       layoutLibrary;
    private Context        context;
    private HomeLibAdapter homeLibAdapter;
    private List<HomeLibraryInfo> homeLibraryInfoArrayList = new ArrayList<>();
    private Theme       theme;
    private int         currentOrientation;
    private boolean     isActionModeActive;
    private ImageButton deleteButton;


    HomeLibrary(HomeUiView homeUiView, Theme theme) {
        this.homeUiView = homeUiView;
        this.context = homeUiView.getContext();
        this.theme = theme;
        init();
    }

    private void setTheme(Theme theme) {
        switch (theme) {
            case DARK:
                layoutLibrary.setCardBackgroundColor(ContextCompat.getColor(context, R.color
                        .dark_home_card_bg));
                break;
            case LIGHT:
                layoutLibrary.setCardBackgroundColor(ContextCompat.getColor(context, R.color
                        .light_home_card_bg));
                break;
        }
    }

    private void init() {
        libraryList = homeUiView.findViewById(R.id.libraryContainer);
        layoutLibrary = homeUiView.findViewById(R.id.cardViewLibrary);
        setTheme(theme);
        deleteButton = homeUiView.findViewById(R.id.deleteButton);
        if (theme == Theme.LIGHT) {
            deleteButton.setImageResource(R.drawable.ic_delete_black);
        } else {
            deleteButton.setImageResource(R.drawable.ic_delete_white);
        }
        currentOrientation = homeUiView.getConfiguration().orientation;
        initList();
        setListeners();
        homeUiView.getLibraries();
    }

    private void initList() {
        libraryList.setItemAnimator(new DefaultItemAnimator());
        libraryList.setHasFixedSize(true);
        libraryList.setNestedScrollingEnabled(false);
        homeLibAdapter = new HomeLibAdapter(context, homeLibraryInfoArrayList, theme);
        homeLibAdapter.setHasStableIds(true);
        libraryList.getItemAnimator().setChangeDuration(0);
        setGridColumns(homeUiView.getConfiguration());

        ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(homeLibAdapter);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);
        itemTouchHelper.attachToRecyclerView(libraryList);
        libraryList.setAdapter(homeLibAdapter);

    }

    private void setListeners() {
        homeLibAdapter.setOnItemClickListener(new HomeLibAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(int position) {
                if (isActionModeActive()) {
                    itemClickActionMode(position, false);
                } else {
                    handleItemClick(position);
                }

            }
        });

        homeLibAdapter.setOnItemLongClickListener(new HomeLibAdapter.OnItemLongClickListener() {
            @Override
            public void onItemLongClick(View view, int position) {
                itemClickActionMode(position, true);
            }
        });

        deleteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SparseBooleanArray sparseBooleanArray = homeLibAdapter.getSelectedItemPositions();
                List<HomeLibraryInfo> fileInfoList = new ArrayList<>();
                for (int i = 0; i < sparseBooleanArray.size(); i++) {
                    fileInfoList.add(homeLibraryInfoArrayList.get(sparseBooleanArray.keyAt(i)));
                }
                homeLibraryInfoArrayList.removeAll(fileInfoList);
                homeLibAdapter.updateAdapter(homeLibraryInfoArrayList);
                endActionMode();
                reloadAftDelete();
            }
        });
    }

    private void reloadAftDelete() {
        List<LibrarySortModel> sortModelList = new ArrayList<>();
        for (int i = 0; i < homeLibraryInfoArrayList.size(); i++) {
            Category category = homeLibraryInfoArrayList.get(i).getCategory();
            if (category.equals(ADD)) {
                continue;
            }
            LibrarySortModel model = new LibrarySortModel();
            model.setChecked(true);
            model.setCategory(category);
            model.setLibraryName(homeLibraryInfoArrayList.get(i).getCategoryName());
            sortModelList.add(model);
        }
        homeUiView.reloadLibs(sortModelList);
    }

    private void handleItemClick(int position) {
        Category category = homeLibraryInfoArrayList.get(position).getCategory();
        if (isAddCategory(category)) {
            Analytics.getLogger().addLibClicked();
            Intent intent = new Intent(context, LibrarySortActivity.class);
            homeUiView.getFragment().startActivityForResult(intent, LIBSORT_REQUEST_CODE);
        } else {
            String path = null;
            if (category.equals(DOWNLOADS)) {
                path = StorageUtils.getDownloadsDirectory();
            }
            homeUiView.loadFileList(path, category);
        }
    }

    private void itemClickActionMode(int position, boolean isLongPress) {
        if (position == homeLibraryInfoArrayList.size() - 1) {
            return;
        }
        homeLibAdapter.toggleSelection(position, isLongPress);

        boolean hasCheckedItems = homeLibAdapter.getSelectedCount() > 0;
        if (hasCheckedItems && !isActionModeActive) {
            startActionMode();
        } else if (!hasCheckedItems && isActionModeActive) {
            endActionMode();
        }
    }

    private void startActionMode() {
        isActionModeActive = true;
        deleteButton.setVisibility(View.VISIBLE);
    }

    private void endActionMode() {
        deleteButton.setVisibility(View.GONE);
        isActionModeActive = false;
        homeLibAdapter.clearSelection();
    }

    private boolean isActionModeActive() {
        return isActionModeActive;
    }


    private void setGridColumns(Configuration configuration) {
        int gridColumns = ConfigurationHelper.getHomeGridCols(configuration);//context.getResources().getInteger(R.integer.homescreen_columns);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(context, gridColumns);
        Log.d(TAG, "setGridColumns: " + gridColumns);
        libraryList.setLayoutManager(gridLayoutManager);
    }

    private void inflateLibraryItem() {
        Log.d(TAG, "inflateLibraryItem: " + homeLibraryInfoArrayList.size()); // TODO: 02/11/17
        // NPE here in dual mode when orientation change from LAND->PORT
        List<String> libNames = new ArrayList<>();
        for (HomeLibraryInfo libraryInfo : homeLibraryInfoArrayList) {
            libNames.add(libraryInfo.getCategoryName());
        }

        homeLibAdapter.updateAdapter(homeLibraryInfoArrayList);
        Analytics.getLogger().homeLibsDisplayed(homeLibraryInfoArrayList.size(), libNames);
    }


    void setLibraries(List<HomeLibraryInfo> libraries) {
        this.homeLibraryInfoArrayList = libraries;
        Log.d(TAG, "setLibraries: " + libraries.size());
        inflateLibraryItem();
    }

    List<HomeLibraryInfo> getLibraries() {
        return homeLibraryInfoArrayList;
    }


    void updateFavoritesCount(int count) {
        for (int i = 0; i < homeLibraryInfoArrayList.size(); i++) {
            if (isFavoritesCategory(homeLibraryInfoArrayList.get(i).getCategory())) {
                homeLibAdapter.updateFavCount(i, count);
                break;
            }
        }
    }

    void onDataLoaded(int id, List<FileInfo> data) {
        for (int i = 0; i < homeLibraryInfoArrayList.size(); i++) {
            int categoryId = homeLibraryInfoArrayList.get(i).getCategory().getValue();
            if (id == categoryId) {
                int count;
                if (id == DOWNLOADS.getValue()) {
                    count = data.size();
                } else {
                    count = data.get(0).getCount();
                }
                homeLibAdapter.updateCount(i, count);
            }
        }
    }


    private boolean isAddCategory(Category category) {
        return category.equals(ADD);
    }

    private boolean isFavoritesCategory(Category category) {
        return category.equals(FAVORITES);
    }

    void onOrientationChanged(Configuration configuration) {
        int orientation = configuration.orientation;
        Log.d(TAG, "onOrientationChanged: old:"+currentOrientation + " neew:"+orientation);
        if (currentOrientation != orientation) {
            currentOrientation = orientation;
            setGridColumns(configuration);
            inflateLibraryItem();
        }
    }

}

File: app/src/main/java/com/siju/acexplorer/home/view/HomeLibrary.kt
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.home.view

import android.content.Context
import android.content.res.Configuration
import android.view.View
import android.widget.ImageButton

import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

import com.siju.acexplorer.R
import com.siju.acexplorer.analytics.Analytics
import com.siju.acexplorer.home.types.HomeLibraryInfo
import com.siju.acexplorer.main.model.StorageUtils
import com.siju.acexplorer.main.model.groups.Category
import com.siju.acexplorer.storage.view.custom.helper.SimpleItemTouchHelperCallback
import com.siju.acexplorer.theme.Theme
import com.siju.acexplorer.utils.ConfigurationHelper

import java.util.ArrayList

import com.siju.acexplorer.main.model.groups.Category.DOWNLOADS
import com.siju.acexplorer.main.model.groups.Category.GENERIC_IMAGES
import com.siju.acexplorer.main.model.groups.Category.GENERIC_MUSIC
import com.siju.acexplorer.main.model.groups.Category.GENERIC_VIDEOS

/**
 * Created by Siju on 03 September,2017
 */
class HomeLibrary(private val context: Context) {

    private val TAG = this.javaClass.simpleName
    private var libraryList: RecyclerView? = null
    private var layoutLibrary: CardView? = null
    private var homeLibAdapter: HomeLibAdapter? = null
    private var homeLibraryInfoArrayList: MutableList<HomeLibraryInfo> = ArrayList()
    private var currentOrientation: Int = 0
    private var isActionModeActive: Boolean = false
    private var deleteButton: ImageButton? = null

    var libraries: MutableList<HomeLibraryInfo>
        get() = homeLibraryInfoArrayList
        set(libraries) {
            this.homeLibraryInfoArrayList = libraries
            inflateLibraryItem()
        }

    init {
        init()
    }

    private fun setTheme(theme: Theme) {
        when (theme) {
            Theme.DARK -> layoutLibrary!!.setCardBackgroundColor(ContextCompat.getColor(context, R.color
                    .dark_home_card_bg))
            Theme.LIGHT -> layoutLibrary!!.setCardBackgroundColor(ContextCompat.getColor(context, R.color
                    .light_home_card_bg))
        }
    }

    private fun init() {
        libraryList = homeUiView.findViewById(R.id.categoryList)
        layoutLibrary = homeUiView.findViewById(R.id.cardViewLibrary)
        setTheme(theme)
        deleteButton = homeUiView.findViewById(R.id.deleteButton)
        if (theme == Theme.LIGHT) {
            deleteButton!!.setImageResource(R.drawable.ic_delete_black)
        } else {
            deleteButton!!.setImageResource(R.drawable.ic_delete_white)
        }
        currentOrientation = homeUiView.configuration.orientation
        initList()
        setListeners()
        homeUiView.getLibraries()
    }

    private fun initList() {
        libraryList!!.itemAnimator = DefaultItemAnimator()
        libraryList!!.setHasFixedSize(true)
        libraryList!!.isNestedScrollingEnabled = false
        homeLibAdapter = HomeLibAdapter(context)
        homeLibAdapter!!.setHasStableIds(false)
        libraryList!!.itemAnimator!!.changeDuration = 0
        setGridColumns(homeUiView.configuration)

        val callback = SimpleItemTouchHelperCallback(homeLibAdapter)
        val itemTouchHelper = ItemTouchHelper(callback)
        itemTouchHelper.attachToRecyclerView(libraryList)
        libraryList!!.adapter = homeLibAdapter
    }

    private fun setListeners() {
        homeLibAdapter!!.setOnItemClickListener(object : HomeLibAdapter.OnItemClickListener {
            override fun onItemClick(position: Int) {
                if (isActionModeActive) {
                    itemClickActionMode(position, false)
                } else {
                    handleItemClick(position)
                }

            }
        })

        homeLibAdapter!!.setOnItemLongClickListener(object : HomeLibAdapter.OnItemLongClickListener {
            override fun onItemLongClick(view: View, position: Int) {
                itemClickActionMode(position, true)
            }
        })

        deleteButton!!.setOnClickListener {
            val sparseBooleanArray = homeLibAdapter!!.getSelectedItemPositions()
            val fileInfoList = ArrayList<HomeLibraryInfo>()
            for (i in 0 until sparseBooleanArray.size()) {
                fileInfoList.add(homeLibraryInfoArrayList[sparseBooleanArray.keyAt(i)])
            }
            homeLibraryInfoArrayList.removeAll(fileInfoList)
            homeLibAdapter!!.updateAdapter(homeLibraryInfoArrayList)
            endActionMode()
            reloadAftDelete()
        }
    }

    private fun reloadAftDelete() {
        homeUiView.reloadLibs(formLibSortList())
    }


    private fun handleItemClick(position: Int) {
        var category = homeLibraryInfoArrayList[position].category
        var path: String? = null
        when (category) {
            DOWNLOADS -> path = StorageUtils.downloadsDirectory
            Category.AUDIO -> category = GENERIC_MUSIC
            Category.IMAGE -> category = GENERIC_IMAGES
            Category.VIDEO -> category = GENERIC_VIDEOS
        }
        homeUiView.loadFileList(path, category)

    }

    private fun itemClickActionMode(position: Int, isLongPress: Boolean) {
        if (position == homeLibraryInfoArrayList.size - 1) {
            return
        }
        homeLibAdapter!!.toggleSelection(position, isLongPress)

        val hasCheckedItems = homeLibAdapter!!.getSelectedCount() > 0
        if (hasCheckedItems && !isActionModeActive) {
            startActionMode()
        } else if (!hasCheckedItems && isActionModeActive) {
            endActionMode()
        }
    }

    private fun startActionMode() {
        isActionModeActive = true
        deleteButton!!.visibility = View.VISIBLE
    }

    private fun endActionMode() {
        deleteButton!!.visibility = View.GONE
        isActionModeActive = false
        homeLibAdapter!!.clearSelection()
        saveLibs()
    }


    private fun setGridColumns(configuration: Configuration) {
        val gridColumns = ConfigurationHelper.getHomeGridCols(configuration)//context.getResources().getInteger(R.integer.homescreen_columns);
        val gridLayoutManager = GridLayoutManager(context, gridColumns)
        libraryList!!.layoutManager = gridLayoutManager
    }

    private fun inflateLibraryItem() {
        // NPE here in dual mode when orientation change from LAND->PORT
        val libNames = ArrayList<String>()
        for (libraryInfo in homeLibraryInfoArrayList) {
            libNames.add(libraryInfo.categoryName)
        }

        homeLibAdapter!!.updateAdapter(homeLibraryInfoArrayList)
        Analytics.getLogger().homeLibsDisplayed(homeLibraryInfoArrayList.size, libNames)
    }


    fun onOrientationChanged(configuration: Configuration) {
        val orientation = configuration.orientation
        if (currentOrientation != orientation) {
            currentOrientation = orientation
            setGridColumns(configuration)
            inflateLibraryItem()
        }
    }

    companion object {
        val LIBSORT_REQUEST_CODE = 1000
    }

}
